lib/screens/v2/candidate_generator/widgets/train_crowding.ex

```diff
@@ -63,36 +88,206 @@ defmodule Screens.V2.CandidateGenerator.Widgets.TrainCrowding do
+        fetch_predictions_fn: fetch_predictions_fn,
+        fetch_params: params,
+        now: now
+      }
+
+      get_instance(
+        any_alert_makes_this_a_terminal?(alerts, location_context),
+        common_params
+      )
+    else
+      :error -> []
+    end
+  end
+
+  defp get_instance(
+         alert_makes_this_a_terminal,
+         common_params
+       ) do
+    next_train_prediction = common_params.next_train_prediction
+
+    cond do
+      is_nil(next_train_prediction) or
+        alert_makes_this_a_terminal or
+          next_train_prediction.vehicle.carriages == [] ->
+        []
+
       # If there is an upcoming train, it's headed to this station, and we're not at a temporary terminal,
       # show the widget
-      if not is_nil(next_train_prediction) and
-           Prediction.vehicle_status(next_train_prediction) in [:in_transit_to, :incoming_at] and
-           next_train_prediction |> Prediction.stop_for_vehicle() |> fetch_parent_stop_id_fn.() ==
-             train_crowding.station_id and
-           next_train_prediction.vehicle.carriages != [] and
-           not any_alert_makes_this_a_terminal?(alerts, location_context) do
-        _ =
-          log_crowding_info(
-            next_train_prediction,
-            logging_options,
-            train_crowding,
-            fetch_predictions_fn,
-            fetch_parent_stop_id_fn,
-            params
-          )
+      Prediction.vehicle_status(next_train_prediction) in [:in_transit_to, :incoming_at] and
+          next_train_prediction
+          |> Prediction.stop_for_vehicle()
+          |> common_params.fetch_parent_stop_id_fn.() ==
+            common_params.train_crowding_config.station_id ->
+        _ = log_crowding_info(:in_transit, common_params)
+
+        Agent.delete(
+          common_params.train_crowding_config.station_id,
+          common_params.train_crowding_config.direction_id
+        )
 
         [
           %CrowdingWidget{
-            screen: config,
+            screen: common_params[:screen_config],
             prediction: next_train_prediction,
-            now: now
+            now: common_params[:now]
           }
         ]
+
+      # Test other heuristics
+      true ->
+        log_heuristics(common_params)
+    end
+  end
+
+  defp log_heuristics(common_params) do
+    train_crowding_config = common_params.train_crowding_config
+    next_train_prediction = common_params.next_train_prediction
+
+    ol_stop_sequence =
+      if train_crowding_config.direction_id == 0 do
+        @ol_station_to_platform_map
       else
-        []
+        Enum.reverse(@ol_station_to_platform_map)
       end
+
+    previous_stop_index =
+      Enum.find_index(
+        ol_stop_sequence,
+        &(elem(&1, 0) == train_crowding_config.station_id)
+      ) - 1
+
+    {_, platform_id_tuple} = Enum.at(ol_stop_sequence, previous_stop_index)
+
+    relevant_platform_id = elem(platform_id_tuple, train_crowding_config.direction_id)
```

Comment:
I tried it out and it gets pretty confusing. The `find_index` function is a little too complicated to map out correctly in a pipe (the formatter was actually not letting me put parentheses where they actually needed to be). The other two lines get confusing because we need two consecutive `elem` calls. I did change the variable name, though.